clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -552,9 +552,11 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
         if (response.wasDisconnected()) {
             log.trace("Cancelled request with header {} due to node {} being disconnected",
                 requestHeader, response.destination());
-            for (ProducerBatch batch : batches.values())
-                completeBatch(batch, new ProduceResponse.PartitionResponse(Errors.NETWORK_EXCEPTION, String.format("Disconnected from node %s", response.destination())),
+            for (ProducerBatch batch : batches.values()) {
+                Errors e = response.wasTimedOut() ? Errors.REQUEST_TIMED_OUT : Errors.NETWORK_EXCEPTION;
+                completeBatch(batch, new ProduceResponse.PartitionResponse(e, String.format("Disconnected from node %s", response.destination())),
```

Comment:
nit: Should we also change the message here when the request is timed out? I also wonder if we should update the trace log at L553 as well. Perhaps, we should just add a `if (response.wasTimeOut)` before `if (response.wasDisconnected())` to fully customise it.

Reply:
I updated the patch to introduce a new `if` block for timeouts that is separate from disconnections, as you'd suggested. LMK if I interpreted your request correctly 😄